plan/rule_table_partition.go

```diff
@@ -0,0 +1,158 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+// // Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package plan
+
+import (
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/ast"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/util/ranger"
+)
+
+// tablePartition rewrites the ast for table partition.
+//
+// create table t (id int) partition by range (id)
+//   (partition
+//      p1 values less than (10),
+//      p2 values less than (20),
+//      p3 values less than (30))
+//
+// select * from t is equal to
+// select * from (union all
+//      select * from p1 where id < 10
+//      select * from p2 where id < 20
+//      select * from p3 where id < 30)
+//
+// tablePartition is here because it's easier to prune partition after predicate push down.
+type tablePartition struct{}
+
+func (s *tablePartition) optimize(lp LogicalPlan) (LogicalPlan, error) {
+	return rewriteDataSource(nil, lp)
+}
+
+func rewriteDataSource(sel *LogicalSelection, lp LogicalPlan) (LogicalPlan, error) {
+	// Assert there will not be sel -> sel in the ast.
+	switch lp.(type) {
+	case *DataSource:
+		return prunePartition(sel, lp.(*DataSource))
+	case *LogicalSelection:
+		return rewriteDataSource(lp.(*LogicalSelection), lp.Children()[0])
+	default:
+		children := lp.Children()
+		for i, child := range children {
+			child1, err := rewriteDataSource(nil, child)
+			if err != nil {
+				return nil, errors.Trace(err)
+			}
+			children[i] = child1
+		}
+	}
+
+	return selectOnSomething(sel, lp)
+}
+
+func selectOnSomething(sel *LogicalSelection, lp LogicalPlan) (LogicalPlan, error) {
```

Comment:
I think we can make all the functions this file to be a member function of `tablePartition`